### PR TITLE
Implement es emitter

### PR DESCRIPTION
**Pull request by itsdawei**
## Description

*This PR is a split from #207*

This PR introduces major API changes for the `Emitter` class. `Emitter` class now requires a `Ranker` object and a `Selector` object.

The `Ranker` object will define a `rank` function which returns the index of the solutions in the descending order that they should be in. It will also define a `reset` function which resets the internal states of the object.

We also introduce `FilterSelector` and `MuSelector` objects which defines a `select()` function that decides how many parents to select from the potential parents.

**Introduced EvolutionStrategyEmitter(x0, sigma0, ranker, restart_rule, bounds, batch_size, seed), which can serve as 
`RandomDirectionEmitter`, `OptimizingEmitter`, and `ImprovementEmitter` depending on the ranker object.**

## Bugs Discovered

- We passed in the same RandomDirectionRanker to the emitters in sphere example, causing the emitter to fail because the ranker is supposed to maintain different random directions.
- RandomDirectionRanker was sorting the solutions using the wrong keys.

## TODO

- [x] Implement `EvolutionStrategyEmitter`
- [x] Pytest
- [x] Test `examples/sphere`
- [x] Sphinx docs
- [x] Fix bugs that we discovered
- [x] Describe bugs discovered (see above)

## Status

- [x] I have read the guidelines in [CONTRIBUTING.md](https://github.com/icaros-usc/pyribs/blob/master/CONTRIBUTING.md).
- [x] I have formatted my code using `yapf`
- [x] I have tested my code by running `pytest`
- [x] I have linted my code with `pylint`
- [x] I have added a one-line description of my change to the changelog in `HISTORY.md`.
- [x] This PR is ready to go

### CODE REVIEW

File: ribs/emitters/rankers.py [hunk 1]
```diff
@@ -303,8 +303,8 @@ def get_ranker(name):
     """Constructs and returns a ranker object based on its string name.
```
Comment (btjanaka) [inline ribs/emitters/rankers.py:303]: change this

File: tests/emitters/rankers_test.py [hunk 2]
```diff
@@ -99,7 +99,8 @@ def test_two_stage_random_direction():
     indices = ranker.rank(emitter, archive, None, solutions, objective_values,
                           behavior_values, metadata, statuses, values)
 
-    assert (indices == [1, 3, 2, 0]).all()
+    print(statuses)
```
Comment (btjanaka) [inline tests/emitters/rankers_test.py:102]: no print

**Comment by itsdawei** [comment]
For reference, we are gonna leave a table that shows the performance of the previous implementation using different emitters and the new implementation that adapts a emitter-ranker structure (using only the evolutionary_strategy_emitter).


                             | New Implementation              | Old Implementation
----------------- | ----------------------------- | --
cma_me_imp       | 91.789%, 17487494 QD, 102s | 91.072%, 17475874 QD, 102s
cma_me_imp_mu | N/A                                           | 92.986%, 17683259 QD, 104s
cma_me_rd          | 90.352%, 14780492 QD, 111s | 92.061%, 12029980 QD, 110s
cma_me_rd_mu   | 96.671%, 11783835 QD, 105s | 95.455%, 11579076 QD, 107s
cma_me_opt        | 70.462%, 14341150 QD, 100s | N/A
cma_me_opt_mu | 12.224%, 2542591 QD, 112s    | 12.495%, 2619242 QD, 96s
cma_me_mixed   | 93.488%, 17449359 QD, 100s | 95.576%, 17841624 QD, 106s

File: HISTORY.md [hunk 3]
```diff
@@ -6,6 +6,7 @@
 
 #### API
 
+- Implement the `evolution_strategy_emitter` (#220)
```
Comment (btjanaka) [inline HISTORY.md:9]: `EvolutionStrategyEmitter`

**Review by btjanaka — APPROVED** [review]

**Review by btjanaka — APPROVED** [review]

**Pull request merged**